MediaBrowser.Providers/Lyric/LrcLyricProvider.cs

```diff
@@ -0,0 +1,200 @@
+using MediaBrowser.Controller.Resolvers;
+using Microsoft.Extensions.Logging;
+
+namespace MediaBrowser.Providers.Lyric;
+
+/// <summary>
+/// LRC Lyric Provider.
+/// </summary>
+public class LrcLyricProvider : ILyricProvider
+{
+    private readonly ILogger<LrcLyricProvider> _logger;
+
+    private readonly LyricParser _lrcLyricParser;
+
+    private static readonly string[] _acceptedTimeFormats = { "HH:mm:ss", "H:mm:ss", "mm:ss", "m:ss" };
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="LrcLyricProvider"/> class.
+    /// </summary>
+    /// <param name="logger">Instance of the <see cref="ILogger"/> interface.</param>
+    public LrcLyricProvider(ILogger<LrcLyricProvider> logger)
+    {
+        _logger = logger;
+        _lrcLyricParser = new LrcParser.Parser.Lrc.LrcParser();
+    }
+
+    /// <inheritdoc />
+    public string Name => "LrcLyricProvider";
+
+    /// <summary>
+    /// Gets the priority.
+    /// </summary>
+    /// <value>The priority.</value>
+    public ResolverPriority Priority => ResolverPriority.First;
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<string> SupportedMediaTypes { get; } = new[] { "lrc", "elrc" };
+
+    /// <summary>
+    /// Opens lyric file for the requested item, and processes it for API return.
+    /// </summary>
+    /// <param name="item">The item to to process.</param>
+    /// <returns>If provider can determine lyrics, returns a <see cref="LyricResponse"/> with or without metadata; otherwise, null.</returns>
+    public LyricResponse? GetLyrics(BaseItem item)
+    {
+        string? lyricFilePath = this.GetLyricFilePath(item.Path);
+
+        if (string.IsNullOrEmpty(lyricFilePath))
+        {
+            return null;
+        }
+
+        var fileMetaData = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+        string lrcFileContent = System.IO.File.ReadAllText(lyricFilePath);
+
+        Song lyricData;
+
+        try
+        {
+            lyricData = _lrcLyricParser.Decode(lrcFileContent);
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Error parsing lyric file {LyricFilePath} from {Provider}", lyricFilePath, Name);
+            return null;
+        }
+
+        List<LrcParser.Model.Lyric> sortedLyricData = lyricData.Lyrics.Where(x => x.TimeTags.Count > 0).OrderBy(x => x.TimeTags.First().Value).ToList();
+
+        // Parse metadata rows
+        var metaDataRows = lyricData.Lyrics
+            .Where(x => x.TimeTags.Count == 0)
+            .Where(x => x.Text.StartsWith('[') && x.Text.EndsWith(']'))
+            .Select(x => x.Text)
+            .ToList();
+
+        foreach (string metaDataRow in metaDataRows)
+        {
+            if (!metaDataRow.Contains(':', StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            string[] metaDataField = metaDataRow.Split(':', 2, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
```

Comment:
Using `IndexOf(':')` and `AsSpan()` with ranges and `Trim()` would be way more efficient
You would only have 2 string allocations (after the span has been checked if it's empty) and only 1 strchr

Reply:
I don't know about way more efficient. It saves some array allocations, but internally Split already uses Spans

Reply:
We already do a `Contains` for a fast path, might as well do everything so we don't have that array alloc

Reply:
I'm not 100% sure how to apply what @Bond-009 is suggesting. Can someone post a suggested edit for this one?

Reply:
The math is probably off, but this is one way
```C#
            var index = metaDataRow.IndexOf(':');
            if (index == -1)
            {
                continue;
            }

            var metaDataFieldName = metaDataRow.AsSpan(1, index - 1).Trim().ToString();
            var metaDataFieldValue = metaDataRow.AsSpan(index + 1, metaDataRow.Length - index).Trim().ToString();
```

Reply:
The analyzer really wants me to use Contains() instead of IndexOf().

![image](https://user-images.githubusercontent.com/3330318/191613754-bc4f7fcb-be0a-439c-b233-134acae9e06c.png)

Reply:
Ah, you're exactly right. I was working top down and stopped once the warning showed.

Reply:
FYI, the final "math" that worked was
```
    var metaDataFieldNameSpan = metaDataRow.AsSpan(1, index - 1).Trim();
    var metaDataFieldValueSpan = metaDataRow.AsSpan(index + 1, metaDataRow.Length - index - 2).Trim();
```